test/common/redis/codec_impl_test.cc

```diff
@@ -0,0 +1,156 @@
+#include "common/redis/codec_impl.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "test/mocks/redis/mocks.h"
+#include "test/test_common/utility.h"
+
+namespace Redis {
+
+class RedisEncoderDecoderImplTest : public testing::Test, public DecoderCallbacks {
+public:
+  RedisEncoderDecoderImplTest() : decoder_(*this) {}
+
+  // Redis::DecoderCallbacks
+  void onRespValue(RespValuePtr&& value) override {
+    decoded_values_.emplace_back(std::move(value));
+  }
+
+  EncoderImpl encoder_;
+  DecoderImpl decoder_;
+  Buffer::OwnedImpl buffer_;
+  std::vector<RespValuePtr> decoded_values_;
+};
+
+TEST_F(RedisEncoderDecoderImplTest, Null) {
+  RespValue value;
+  encoder_.encode(value, buffer_);
+  EXPECT_EQ("$-1\r\n", TestUtility::bufferToString(buffer_));
+  decoder_.decode(buffer_);
+  EXPECT_EQ(value, *decoded_values_[0]);
+  EXPECT_EQ(0UL, buffer_.length());
+}
+
+TEST_F(RedisEncoderDecoderImplTest, Error) {
+  RespValue value;
+  value.type(RespType::Error);
+  value.asString() = "error";
+  encoder_.encode(value, buffer_);
+  EXPECT_EQ("-error\r\n", TestUtility::bufferToString(buffer_));
+  decoder_.decode(buffer_);
+  EXPECT_EQ(value, *decoded_values_[0]);
+  EXPECT_EQ(0UL, buffer_.length());
+}
+
+TEST_F(RedisEncoderDecoderImplTest, SimpleString) {
+  RespValue value;
+  value.type(RespType::SimpleString);
+  value.asString() = "simple string";
+  encoder_.encode(value, buffer_);
+  EXPECT_EQ("+simple string\r\n", TestUtility::bufferToString(buffer_));
+  decoder_.decode(buffer_);
+  EXPECT_EQ(value, *decoded_values_[0]);
+  EXPECT_EQ(0UL, buffer_.length());
+}
+
+TEST_F(RedisEncoderDecoderImplTest, Integer) {
```

Comment:
I'm not really sure what other clients do about overflow in general. I will add a follow up on the GH issue to figure out how to deal with overflow in the codec.